File: src/features/toggles/toggleHooks.ts
import { useCallback, useEffect, useState } from "react";

import { getStoredToggle } from "./toggleUtils";
import { useStorageListener } from "../chrome/chromeHooks";

export function useToggle(name: string): boolean {
  const [toggle, setToggle] = useState<boolean>(false);

  const populateToggle = useCallback(() => {
    getStoredToggle(name).then((value) => {
      setToggle(value);
    });
  }, [name]);

  useEffect(() => {
    populateToggle();
  }, [populateToggle]);

  useStorageListener(populateToggle);

  return toggle;
}

File: src/main/kotlin/features/toggles/toggleHooks.kt
package features.toggles

import dom.html.HTMLInputElement
import features.chrome.setStoredValue
import react.dom.events.ChangeEvent
import react.useEffectOnce
import react.useState

fun useToggle(name: String): Pair<Boolean, (event: ChangeEvent<HTMLInputElement>, checked: Boolean) -> Unit> {
    val (toggle, setToggle) = useState(false)

    useEffectOnce {
        getStoredToggle(name).then { setToggle(it) }
    }

    return Pair(toggle) { event, _ ->
        setToggle(event.target.checked)
        setStoredValue(event.target.id, event.target.checked)
    }
}
